### PR TITLE
Toriesz

### PR DESCRIPTION
**Pull request by rckirby**
This PR provides a more general (and efficient) to_riesz method for a dual set that aggregates all the tabulation of functionals and derivatives across all the functionals in the set.  This bypasses special-purpose code (that outsources to sympy) in subclasses of Functional.  The result is that we can delete a lot of code and get a nice speedup for instantiation of Argyris and other elements with derivative dofs.

The logic that goes in will also be useful later in automating dual_evaluation up in FInAT.

### CODE REVIEW

**Comment by colinjcotter** [comment]
Unfortunate name...

On 14 Jan 2019, at 20:59, Robert Kirby <notifications@github.com<mailto:notifications@github.com>> wrote:


This PR provides a more general (and efficient) to_riesz method for a dual set that aggregates all the tabulation of functionals and derivatives across all the functionals in the set. This bypasses special-purpose code (that outsources to sympy) in subclasses of Functional. The result is that we can delete a lot of code and get a nice speedup for instantiation of Argyris and other elements with derivative dofs.

The logic that goes in will also be useful later in automating dual_evaluation up in FInAT.

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/firedrakeproject/fiat/pull/12

Commit Summary

  *   general/abstract to_riesz, lift to dual
  *   Add WuXu H3 nonconforming element (first stab)
  *   Save WuXu element for later
  *   Switch to new to_riesz method

File Changes

  *   M FIAT/dual_set.py<https://github.com/firedrakeproject/fiat/pull/12/files#diff-0> (81)
  *   M FIAT/functional.py<https://github.com/firedrakeproject/fiat/pull/12/files#diff-1> (116)

Patch Links:

  *   https://github.com/firedrakeproject/fiat/pull/12.patch
  *   https://github.com/firedrakeproject/fiat/pull/12.diff

File: FIAT/dual_set.py [hunk 1]
```diff
@@ -49,18 +51,91 @@ def get_entity_ids(self):
...
+
+    #     riesz_shape = tuple([num_nodes] + list(tshape) + [num_exp])
+
+    #     self.mat = numpy.zeros(riesz_shape, "d")
+
+    #     for i in range(len(self.nodes)):
+    #         self.mat[i][:] = self.nodes[i].to_riesz(poly_set)
+
+    #     return self.mat
+
```
Comment (thomasgibson) [inline FIAT/dual_set.py]: commented out code should be removed

File: FIAT/functional.py [hunk 2]
```diff
@@ -123,7 +129,25 @@ def to_riesz(self, poly_set):
...
+
+            ndpts = len(dpts)
+            for j in range(ndpts):
+                dpt_cur = dpts[j]
+                wac_list = dpt_dict[dpt_cur]
+                for i in range(nbf):
+                    for (w, alpha, c) in wac_list:
+                        result[c][i] += w * dbfs[alpha][i, j]
+
+            #raise NotImplementedError("Generic to_riesz implementation does not support derivatives")
```
Comment (thomasgibson) [inline FIAT/functional.py]: left over debugging or intended use elsewhere?

File: FIAT/functional.py [hunk 3]
```diff
@@ -191,24 +215,24 @@ def __call__(self, fn):
...
-                      for count in range(a))
+    #     # Expand the multi-index as a series of variables to
+    #     # differentiate with respect to.
+    #     dvars = tuple(d for d, a in zip(dx, self.alpha)
+    #                   for count in range(a))
 
-        return numpy.asarray([sympy.lambdify(X, sympy.diff(b, *dvars))(x)
-                              for b in bfs])
+    #     return numpy.asarray([sympy.lambdify(X, sympy.diff(b, *dvars))(x)
+    #                           for b in bfs])
```
Comment (thomasgibson) [inline FIAT/functional.py]: commented out code should be removed

File: FIAT/functional.py [hunk 4]
```diff
@@ -223,11 +247,55 @@ def __init__(self, ref_el, facet_no, pt):
...
+    #     ed = poly_set.get_embedded_degree()
+
+    #     bfs = es.tabulate(ed, [dx])[:, 0]
+
+    #     # We need the gradient dotted with the normal.
+    #     return numpy.asarray(
+    #         [sympy.lambdify(
+    #             X, sum([sympy.diff(b, dxi)*ni
+    #                     for dxi, ni in zip(dx, self.n)]))(x)
+    #          for b in bfs])
```
Comment (thomasgibson) [inline FIAT/functional.py]: same as above

File: FIAT/dual_set.py [hunk 5]
```diff
@@ -15,7 +15,9 @@
 # You should have received a copy of the GNU Lesser General Public License
 # along with FIAT. If not, see <http://www.gnu.org/licenses/>.
 
-import numpy
+import numpy, collections
```
[the remark this answers is not in the document]
Comment (rckirby) [reply FIAT/dual_set.py]: Changed (I had fixed it but forgotten to push).

File: FIAT/dual_set.py (hunk 1 above)
Comment (rckirby) [reply FIAT/dual_set.py]: Done

File: FIAT/functional.py (hunk 2 above)
Comment (rckirby) [reply FIAT/functional.py]: Thanks for taking a look.  I changed this as we..

File: FIAT/functional.py (hunk 3 above)
Comment (rckirby) [reply FIAT/functional.py]: Done

File: FIAT/dual_set.py [hunk 6]
```diff
@@ -54,13 +57,67 @@ def to_riesz(self, poly_set):
...
         num_exp = es.get_num_members(poly_set.get_embedded_degree())
+        nbf = poly_set.get_num_members()
 
         riesz_shape = tuple([num_nodes] + list(tshape) + [num_exp])
 
         self.mat = numpy.zeros(riesz_shape, "d")
 
-        for i in range(len(self.nodes)):
-            self.mat[i][:] = self.nodes[i].to_riesz(poly_set)
+        # let's amalgamate the pt_dict and deriv_dicts of all the
```
Comment (dham) [inline FIAT/dual_set.py]: Maybe I'm just to dim for this, but I find this code unintelligible. I think a key problem is that the local variable names are so abbreviated that I can't understand what is what.

File: FIAT/dual_set.py [hunk 7]
```diff
@@ -54,13 +57,67 @@ def to_riesz(self, poly_set):
         tshape = self.nodes[0].target_shape
         num_nodes = len(self.nodes)
         es = poly_set.get_expansion_set()
+        ed = poly_set.get_embedded_degree()
         num_exp = es.get_num_members(poly_set.get_embedded_degree())
+        nbf = poly_set.get_num_members()
 
         riesz_shape = tuple([num_nodes] + list(tshape) + [num_exp])
 
         self.mat = numpy.zeros(riesz_shape, "d")
```
Comment (dham) [inline FIAT/dual_set.py]: I realise this is inherited code, but `mat` is a datatype, it's not really a great name for a member.

File: FIAT/dual_set.py [hunk 8]
```diff
@@ -54,13 +57,67 @@ def to_riesz(self, poly_set):
...
 
-        for i in range(len(self.nodes)):
-            self.mat[i][:] = self.nodes[i].to_riesz(poly_set)
+        # let's amalgamate the pt_dict and deriv_dicts of all the
+        # functionals so we can tabulate the basis functions twice only
+        # (once on pts and once on derivatives)
+
+        # Need: dictionary mapping pts to which functionals they come from
+        pts_to_ells = collections.OrderedDict()
```
Comment (dham) [inline FIAT/dual_set.py]: What's an `ell`?

File: FIAT/dual_set.py [hunk 9]
```diff
@@ -54,13 +57,67 @@ def to_riesz(self, poly_set):
...
+        for i, ell in enumerate(self.nodes):
+            for pt in ell.deriv_dict:
+                if pt in dpts_to_ells:
+                    dpts_to_ells[pt].append(i)
+                else:
+                    dpts_to_ells[pt] = [i]
+
+        # Now tabulate
+        pts = list(pts_to_ells.keys())
+        bfs = es.tabulate(ed, pts)
```
Comment (dham) [inline FIAT/dual_set.py]: `bfs`?

File: FIAT/dual_set.py [hunk 10]
```diff
@@ -54,13 +57,67 @@ def to_riesz(self, poly_set):
...
+
+                for i in range(bfs.shape[0]):
+                    for (w, c) in wc_list:
+
+                        self.mat[k][c][i] += w*bfs[i, j]
+
+        mdo = max([ell.max_deriv_order for ell in self.nodes])
+        if mdo > 0:
+            dpts = list(dpts_to_ells.keys())
+            es_foo = polynomial_set.ONPolynomialSet(self.ref_el, ed)
```
Comment (dham) [inline FIAT/dual_set.py]: `ess_foo`?

File: FIAT/dual_set.py [hunk 11]
```diff
@@ -54,13 +57,67 @@ def to_riesz(self, poly_set):
...
+            for k in which_ells:
+                pt_dict = self.nodes[k].pt_dict
+                wc_list = pt_dict[pt]
+
+                for i in range(bfs.shape[0]):
+                    for (w, c) in wc_list:
+
+                        self.mat[k][c][i] += w*bfs[i, j]
+
+        mdo = max([ell.max_deriv_order for ell in self.nodes])
```
Comment (dham) [inline FIAT/dual_set.py]: expand `mdo` please.

File: FIAT/dual_set.py (hunk 7 above)
Comment (rckirby) [reply FIAT/dual_set.py]: Will change to self.matrix.  It only appears in this file.

File: FIAT/dual_set.py (hunk 8 above)
Comment (rckirby) [reply FIAT/dual_set.py]: ell is the letter of the alphabet before em.  \ell is frequently an abbreviation for linear functionals.

File: FIAT/dual_set.py (hunk 9 above)
Comment (rckirby) [reply FIAT/dual_set.py]: basis functions.  Expanded to "expansion_values"

File: FIAT/dual_set.py (hunk 11 above)
Comment (rckirby) [reply FIAT/dual_set.py]: --> max_deriv_order

File: FIAT/dual_set.py (hunk 10 above)
Comment (rckirby) [reply FIAT/dual_set.py]: Temporary expansion set used only here and distinct from the one of the overarching polynomial set.  Will change name and add comments.

**Comment by rckirby** [comment]
Don't merge yet, I'm finding a few new bugs trying to use this!

File: FIAT/dual_set.py (hunk 6 above)
Comment (rckirby) [reply FIAT/dual_set.py]: I've added some top-level documentation explaining a bit about what it does.

**Comment by rckirby** [comment]
Looks like everything is passing.  Can @thomasgibson and @dham and anyone else interested take a look to see if the concerns are addressed?

**Comment by wence-** [comment]
OK, let's do discussion upstream: https://bitbucket.org/fenics-project/fiat/pull-requests/54/improve-implementation-of-to_riesz-method/diff

Thanks!

File: FIAT/dual_set.py [hunk 12]
```diff
@@ -49,18 +51,91 @@ def get_entity_ids(self):
...
+        # Need: dictionary mapping pts to which functionals they come from
+        pts_to_ells = collections.OrderedDict()
+        dpts_to_ells = collections.OrderedDict()
+
+        for i, ell in enumerate(self.nodes):
+            for pt in ell.pt_dict:
+                if pt in pts_to_ells:
+                    pts_to_ells[pt].append(i)
+                else:
+                    pts_to_ells[pt] = [i]
```
[the remark this answers is not in the document]
Comment (miklos1) [reply FIAT/dual_set.py]: Ah, I see...

**Pull request closed without merging**